js/gestionPresupuesto.js

```diff
@@ -1,18 +1,163 @@
-// TODO: Crear las funciones, objetos y variables indicadas en el enunciado
+// Variables globales
+let presupuesto = 0;
+let gastos = [];
+let idGasto = 0;
 
-// TODO: Variable global
+// MIO
+function actualizarPresupuesto(num) {
+    if(num >= 0){
+        presupuesto = num;
+        return presupuesto;
+    }
+    else{
+        console.log("No se puede introducir un número negativo.");
+        return -1;
+    }
+}
 
 
-function actualizarPresupuesto() {
-    // TODO
+// MIO
+function mostrarPresupuesto() {
+    return `Tu presupuesto actual es de ${presupuesto} €`;
 }
 
-function mostrarPresupuesto() {
-    // TODO
+function CrearGasto(descripcion, valor, fecha, ...etiquetas) {
+
+    // Métodos
+    // MIO
+    this.mostrarGasto = function(){
+        return `Gasto correspondiente a ${this.descripcion} con valor ${this.valor} €`;
+    }
+
+    // MIO
+    this.actualizarDescripcion = function(descripcion){
+        this.descripcion = descripcion;
+        return descripcion;
+    }
+
+    // MIO
+    this.actualizarValor = function(valor){
+        if(valor >= 0){
+            this.valor = valor;
+            return valor;
+        }
+        else{
+            valor = this.valor;
+            return valor;
+        }
+    }
+
+    // MIO
+    this.mostrarGastoCompleto = function(){
+        let texto = `Gasto correspondiente a ${this.descripcion} con valor ${this.valor} €.
+Fecha: ${new Date(this.fecha).toLocaleString()}
+Etiquetas:\n`
+            for(let e of this.etiquetas){
+                texto += `- ${e}\n`
+            }
+            return texto;
+    }
+
+    this.actualizarFecha = function(fecha) {
+        let f = Date.parse(fecha);
+        if (f) {
+            this.fecha = f;
+        }
+    }
+
+    // MIO
+    this.anyadirEtiquetas = function(...etiqueta){
+        for(let e of etiqueta){
+            if(this.etiquetas.indexOf(e) == -1){
+                this.etiquetas.push(e);
+            }
+        }
+    }
+    // SOLUCION
+    this.borrarEtiquetas = function(...etqs) {
+        // No entiendo porqué se pone push si se supone que hay que borrar la etiqueta que
+        // coincida con la que nos dan.
+        let newetiquetas = [];
+
+        for (let e of this.etiquetas) {
+	    if (etqs.indexOf(e) == -1) {
+                newetiquetas.push(e);
+	    }
+        }
+
+        this.etiquetas = newetiquetas;
+
+        // No entiendo porqué no funciona el código de abajo.
```

Comment:
Según la documentación (https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Array/slice):
> El método slice() devuelve una copia de una parte del array dentro de un nuevo array empezando por inicio hasta fin (fin no incluido). El array original no se modificará.

Por tanto, `this.etiquetas.slice(f)` no modifica `this.etiquetas`.

Además, `slice` devuelve un trozo del array original, en tu caso indicando solo el inicio. Y el parámetro de `slice` debe ser un número, no un objeto (en tu caso le estás pasando la etiqueta, no la posición donde está la etiqueta).

Quizá en vez de `slice` puedas utilizar `splice`: https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Array/splice